src/util/invariant.h

```diff
@@ -500,9 +500,33 @@ CBMC_NORETURN void report_invariant_failure(
   EXPAND_MACRO(INVARIANT_STRUCTURED(CONDITION, TYPENAME, __VA_ARGS__))
 
 /// This should be used to mark dead code
-#define UNREACHABLE INVARIANT(false, "Unreachable")
-#define UNREACHABLE_STRUCTURED(TYPENAME, ...)                                  \
-  EXPAND_MACRO(INVARIANT_STRUCTURED(false, TYPENAME, __VA_ARGS__))
+#ifdef __GNUC__
+// GCC 12 with -O0 fails reporting missing return when using UNREACHABLE.
+// Using __builtin_unreachable fixes this without breaking the invariant.
+#  define UNREACHABLE                                                          \
+    do                                                                         \
+    {                                                                          \
+      INVARIANT(false, "Unreachable");                                         \
+      __builtin_unreachable();                                                 \
```

Comment:
Shouldn't we instead be adding this to the expansion of the `INVARIANT` macro?

Reply:
No as the invariant can actually be satisfied (`INVARIANT(condition, "some reason")` succeed when `condition` is `true`).

The problem is just related to GCC 12 with `-O0` refusing to statically resolve cases of the form `if (true) { <terminate> }`.

The only cases I found of this are `UNREACHABLE`, `UNIMPLEMENTED_FEATURE` and some `INVARIANT(false, "unreachable because")` and they have all been fixed.

Reply:
What I meant was putting the `__builtin_unreachable()` inside the `if(!(CONDITION))` branch, which wouldn't have the problem of the invariant being satisfied. Wouldn't that have worked? (It's possible that GCC's reachability analysis will remain confused in that case.)